File: src/poly.rs
//! Polynomial helpers shared by Shamir, the multi-secret extension,
//! the ramp scheme, and the errors-and-erasures decoder.
//!
//! Polynomials are stored as a coefficient vector
//! `c = [c_0, c_1, …, c_{d}]` representing
//! `c_0 + c_1 x + c_2 x^2 + … + c_d x^d`.

use crate::field::PrimeField;
use crate::bigint::BigUint;

/// Evaluate `f(x) = c_0 + c_1 x + … + c_{d} x^d` by Horner's rule.
#[must_use]
pub fn horner(field: &PrimeField, coeffs: &[BigUint], x: &BigUint) -> BigUint {
    let mut acc = BigUint::zero();
    for c in coeffs.iter().rev() {
        acc = field.add(&field.mul(&acc, x), c);
    }
    acc
}

/// Lagrange-evaluate the unique polynomial of degree `< points.len()`
/// passing through `points` at the abscissa `x_eval`.
///
/// Returns `None` if any two `x` coordinates collide (which makes the
/// system singular). Shamir 1979 §2 reduces to `lagrange_eval(.., 0)`.
#[must_use]
pub fn lagrange_eval(
    field: &PrimeField,
    points: &[(BigUint, BigUint)],
    x_eval: &BigUint,
) -> Option<BigUint> {
    let n = points.len();
    // Reduce every label modulo p before checking pairwise distinctness
    // — otherwise `x = 1` and `x = p + 1` look distinct but collide
    // inside the field, producing a zero denominator and a panic in
    // `lagrange_eval_unchecked`.
    let reduced: Vec<BigUint> = points.iter().map(|(x, _)| field.reduce(x)).collect();
    for r in &reduced {
        if r.is_zero() {
            // x = 0 is reserved for the secret; refuse honestly.
            return None;
        }
    }
    for i in 0..n {
        for j in (i + 1)..n {
            if reduced[i] == reduced[j] {
                return None;
            }
        }
    }
    Some(lagrange_eval_unchecked(field, points, x_eval))
}

/// Inner Lagrange evaluator that assumes the caller has already
/// verified all `x` coordinates are distinct. Hot-path use only.
///
/// The per-point denominator inverses come from Montgomery's batch
/// trick: invert the single product `∏ den_j` with one extended-gcd
/// call, then peel off each individual inverse with two multiplies.
/// Inversion (several `div_rem` rounds inside `mod_inverse`) dominates
/// the cost of a reconstruction, so doing it once instead of k times
/// is the difference that matters as the threshold grows.
///
/// # Panics
/// Panics if two `x` coordinates collide. Use [`lagrange_eval`] from
/// any path that has not already validated input distinctness.
#[must_use]
pub fn lagrange_eval_unchecked(
    field: &PrimeField,
    points: &[(BigUint, BigUint)],
    x_eval: &BigUint,
) -> BigUint {
    let n = points.len();
    if n == 0 {
        return BigUint::zero();
    }

    // den[j] = ∏_{i≠j} (x_j − x_i); each factor is nonzero in a prime
    // field because the x's are distinct, so every den is invertible.
    let mut dens: Vec<BigUint> = Vec::with_capacity(n);
    for j in 0..n {
        let xj = &points[j].0;
        let mut den = BigUint::one();
        for (i, (xi, _)) in points.iter().enumerate() {
            if i == j {
                continue;
            }
            den = field.mul(&den, &field.sub(xj, xi));
        }
        dens.push(den);
    }

    // Batch inversion. Forward pass: prefix[j] = den_0 · … · den_{j−1}.
    let mut prefix: Vec<BigUint> = Vec::with_capacity(n);
    let mut acc = BigUint::one();
    for den in &dens {
        prefix.push(acc.clone());
        acc = field.mul(&acc, den);
    }
    // Backward pass: peel den_j off the running inverse so that
    // inv_acc always equals (den_0 · … · den_j)^{-1} entering step j.
    let mut inv_acc = field
        .inv(&acc)
        .expect("Lagrange denominator product nonzero given distinct x");
    let mut den_invs = vec![BigUint::zero(); n];
    for j in (0..n).rev() {
        den_invs[j] = field.mul(&inv_acc, &prefix[j]);
        inv_acc = field.mul(&inv_acc, &dens[j]);
    }

    let mut sum = BigUint::zero();
    for j in 0..n {
        let (_, yj) = &points[j];
        let mut num = BigUint::one();
        for (i, (xi, _)) in points.iter().enumerate() {
            if i == j {
                continue;
            }
            // L_j(x) = ∏_{i ≠ j} (x − x_i) / (x_j − x_i)
            num = field.mul(&num, &field.sub(x_eval, xi));
        }
        let term = field.mul(yj, &field.mul(&num, &den_invs[j]));
        sum = field.add(&sum, &term);
    }
    sum
}

#[cfg(test)]
mod tests {
    use super::*;

    fn f257() -> PrimeField {
        PrimeField::new(BigUint::from_u64(257))
    }

    #[test]
    fn horner_matches_manual_eval() {
        // f(x) = 5 + 3x + 2x^2 → f(4) = 5 + 12 + 32 = 49
        let f = f257();
        let coeffs = vec![
            BigUint::from_u64(5),
            BigUint::from_u64(3),
            BigUint::from_u64(2),
        ];
        let v = horner(&f, &coeffs, &BigUint::from_u64(4));
        assert_eq!(v, BigUint::from_u64(49));
    }

    #[test]
    fn lagrange_recovers_polynomial() {
        // f(x) = 7 + 11x + 5x^2; sample at x = 1, 2, 3.
        let f = f257();
        let coeffs = vec![
            BigUint::from_u64(7),
            BigUint::from_u64(11),
            BigUint::from_u64(5),
        ];
        let pts: Vec<(BigUint, BigUint)> = (1..=3)
            .map(|i| {
                let x = BigUint::from_u64(i);
                let y = horner(&f, &coeffs, &x);
                (x, y)
            })
            .collect();
        // Constant term = secret = 7.
        assert_eq!(
            lagrange_eval(&f, &pts, &BigUint::zero()),
            Some(BigUint::from_u64(7))
        );
        // Should also reproduce f(4) = 7 + 44 + 80 = 131.
        assert_eq!(
            lagrange_eval(&f, &pts, &BigUint::from_u64(4)),
            Some(BigUint::from_u64(131))
        );
    }

    #[test]
    fn lagrange_rejects_duplicate_x() {
        let f = f257();
        let pts = vec![
            (BigUint::from_u64(1), BigUint::from_u64(7)),
            (BigUint::from_u64(1), BigUint::from_u64(8)),
        ];
        assert!(lagrange_eval(&f, &pts, &BigUint::zero()).is_none());
    }
}

File: cpp/src/poly.cpp
#include "secret_sharing/poly.hpp"

namespace secret_sharing {

big_uint horner(prime_field const& f, std::span<big_uint const> coefficients,
                big_uint const& x) {
    if (coefficients.empty()) {
        return big_uint::zero();
    }
    auto acc = coefficients.back();
    for (std::size_t i = coefficients.size() - 1; i-- > 0;) {
        acc = f.mul(acc, x);
        acc = f.add(acc, coefficients[i]);
    }
    return acc;
}

std::optional<big_uint> lagrange_eval(prime_field const& f,
                                      std::span<std::pair<big_uint, big_uint> const> points,
                                      big_uint const& target) {
    if (points.empty()) {
        return std::nullopt;
    }
    auto const n = points.size();

    // den[i] = ∏_{j≠i} (x_i − x_j). A zero factor means two abscissae
    // collide (possibly only after reduction mod p), which makes the
    // interpolation system singular — report that instead of dividing
    // by zero.
    std::vector<big_uint> dens;
    dens.reserve(n);
    for (std::size_t i = 0; i < n; ++i) {
        auto const& xi = points[i].first;
        big_uint den = big_uint::one();
        for (std::size_t j = 0; j < n; ++j) {
            if (j == i) {
                continue;
            }
            auto diff = f.sub(xi, points[j].first);
            if (diff.is_zero()) {
                return std::nullopt;  // duplicate abscissa
            }
            den = f.mul(den, diff);
        }
        dens.push_back(std::move(den));
    }

    // Montgomery batch inversion: invert the single product of all
    // denominators with one extended-gcd call, then peel off each
    // individual inverse with two multiplies. Inversion dominates the
    // cost of a reconstruction, so this is the difference between one
    // inversion and k of them. Forward pass builds the prefix
    // products; the backward pass keeps inv_acc equal to
    // (den_0 · … · den_i)⁻¹ entering step i.
    std::vector<big_uint> prefix;  // prefix[i] = den_0 · … · den_{i−1}
    prefix.reserve(n);
    big_uint acc = big_uint::one();
    for (auto const& den : dens) {
        prefix.push_back(acc);
        acc = f.mul(acc, den);
    }
    auto product_inv = f.inv(acc);
    if (!product_inv) {
        return std::nullopt;  // unreachable: every den is nonzero
    }
    big_uint inv_acc = std::move(*product_inv);
    std::vector<big_uint> den_invs(n);
    for (std::size_t i = n; i-- > 0;) {
        den_invs[i] = f.mul(inv_acc, prefix[i]);
        inv_acc = f.mul(inv_acc, dens[i]);
    }

    big_uint result = big_uint::zero();
    for (std::size_t i = 0; i < n; ++i) {
        auto const& yi = points[i].second;
        big_uint num = big_uint::one();
        for (std::size_t j = 0; j < n; ++j) {
            if (j == i) {
                continue;
            }
            // L_i(x) = ∏_{j≠i} (x − x_j) / (x_i − x_j)
            num = f.mul(num, f.sub(target, points[j].first));
        }
        result = f.add(result, f.mul(yi, f.mul(num, den_invs[i])));
    }
    return result;
}

}  // namespace secret_sharing
